closed/src/java.base/share/classes/com/sun/crypto/provider/NativeChaCha20Cipher.java

```diff
@@ -87,20 +89,48 @@
     // The underlying engine for doing the ChaCha20/Poly1305 work
     private ChaChaEngine engine;
 
-    private static NativeCrypto nativeCrypto;
+    private static final NativeCrypto nativeCrypto;
+    private static final Cleaner contextCleaner;
     private long context;
 
     private final ByteArrayOutputStream aadBuf;
 
     static {
         nativeCrypto = NativeCrypto.getNativeCrypto();
+        contextCleaner = CleanerFactory.cleaner();
+    }
+
+    private static final class ChaCha20CleanerRunnable implements Runnable {
+        private final long context;
+
+        public ChaCha20CleanerRunnable(long context) {
+            this.context = context;
+        }
+
+        @Override
+        public void run() {
+            /*
+             * Release the ChaCha20 context.
+             */
+            long ret = nativeCrypto.DestroyContext(context);
+            if (ret == -1) {
+                throw new ProviderException("Error in NativeChaCha20Cipher - DestroyContext");
+            }
+        }
     }
 
     /**
      * Default constructor.
      */
     protected NativeChaCha20Cipher() {
         context = nativeCrypto.CreateContext();
+
+        if (context != -1) {
+            contextCleaner.register(this, new ChaCha20CleanerRunnable(this.context));
+        } else {
+            throw new ProviderException("Error in NativeChaCha20Cipher - CreateContext");
+        }
```

Comment:
Please invert this so `else` is not needed:
```
        if (context == -1) {
            throw new ProviderException("Error in NativeChaCha20Cipher - CreateContext");
        }
        contextCleaner.register(this, new ChaCha20CleanerRunnable(context));
```

Reply:
done